Add Windows support and first-class Linux support

matheuss opened the pull request:
<img src="https://cloud.githubusercontent.com/assets/4721750/19839554/389c61d6-9ecb-11e6-9ed4-a8c6a908437e.gif" width="300px"/>

## TODO
- [x] ~~Fix [`npm prune; npm install --production`](https://github.com/zeit/hyper/blob/master/app/plugins.js#L239) on Windows~~ – https://github.com/zeit/hyper/pull/946/commits/35df44448d0a057a4ded7b536fac72b446e85860
- [x] ~~Fix `NODE_ENV=production webpack` on Windows~~ https://github.com/zeit/hyper/pull/946/commits/a1aacdb7fa9579ff142168479fd36f8a48beff2f
- [x] ~~Add a bigger `icon.ico` for Windows~~ https://github.com/zeit/hyper/pull/946/commits/90287264aee1a647bbab41477b8ea5d8fa34ec4d
- [x] ~Build on AppVeyor~
- [x] ~Add `Building on Windows` to README.md~
- [ ] Add a [GIF to show when installing on Windows](https://github.com/electron-userland/electron-builder/wiki/Options#buildsquirrelwindows) – cc @rauchg @evilrabbit
- [ ] Find out why hotkeys like `Ctrl+T`, `Ctrl+N` etc aren't working – related: #274
- [x] ~Fix the `maximize` and `close` SVGs  – cc @CodeTheory @evilrabbit~
- [x] ~Add padding to the hamburger menu, minimize, maximize and close buttons – cc @CodeTheory @evilrabbit~
- [x] ~Design review – cc @CodeTheory @evilrabbit @rauchg~
- [x] ~Add CSS classes to enable customization of the hamburger menu, minimize, maximize and close buttons~ `.header_shape` is enough
- [ ] Performance review – cc @rauchg
- [x] ~Test on macOS to make sure that `pty.js` doesn't introduce any weird behavior~

## Summary

- `child_pty` => `pty.js` to enable **Windows support** (https://github.com/zeit/hyper/commit/7f29ead1d52d6ad79cf1839d51fba92120419ba6)
- A brand new interface for Linux and Windows _(macOS stays untouched)_ (https://github.com/zeit/hyper/commit/9dbff528390c92ee8bd07b3a99b1724ea9cd7ee3):

![screen shot 2016-10-30 at 8 34 15 pm](https://cloud.githubusercontent.com/assets/4721750/19840685/c62c96e4-9ee2-11e6-9783-b87f391d4a3d.jpg)
![screen shot 2016-10-30 at 8 35 11 pm](https://cloud.githubusercontent.com/assets/4721750/19840686/c96156d8-9ee2-11e6-9b5e-f35ad5147886.jpg)
## Why this new design 🤔

When you create a [frameless window](https://github.com/electron/electron/blob/master/docs/api/frameless-window.md), you need to specify `-webkit-app-region: drag` in CSS to tell Electron which regions are draggable. 
Electron's documentation [states](https://github.com/electron/electron/blob/master/docs/api/frameless-window.md#draggable-region): 

> Note that if you have made the whole window draggable, you must also mark buttons as non-draggable, otherwise it would be impossible for users to click on them

The problem is that on macOS, the above is simply not true. An element specified as the draggable region will receive click events:

![kapture 2016-10-28 at 16 56 30](https://cloud.githubusercontent.com/assets/4721750/19841207/6ccf13c2-9eee-11e6-9f63-1b24d921613f.gif)

On Linux and Windows, the behavior is the one described in the documentation – the element will not receive click events:

![kapture 2016-10-28 at 2 07 50](https://cloud.githubusercontent.com/assets/4721750/19841284/61ebc32c-9ef0-11e6-9899-988486d53087.gif)
_(On Windows/Linux you can drag the window too, I just didn't recorded it)_

With that in mind, we introduced a new `<div>` that becomes visible on Windows/Linux when there are two or more open tabs:

![kapture 2016-10-30 at 22 47 59](https://cloud.githubusercontent.com/assets/4721750/19841448/08d051c4-9ef3-11e6-9d22-083bfcdea7e1.gif)

## Building on Windows

To build on Windows you need to install some dependencies and run some commands: 

1. Install [VC++ Build Tools Technical Preview](http://go.microsoft.com/fwlink/?LinkId=691126) using the **Default Install option**
2. Install Python 2.7, add it to your PATH and run `npm config set python python2.7`
3. Run `npm config set msvs_version 2015 --global`

## About `pty.js`

We're using @Tyriar's [fork](https://github.com/Tyriar/pty.js) of `pty.js` – [the same used](https://github.com/Microsoft/vscode/blob/57fc08a2c9070d67ab42acd4c083e11d9702aed4/package.json#L33) on [Visual Studio Code](https://github.com/Microsoft/vscode). Check #190 for a discussion on that.

## Feeback wanted

We need your feedback to make sure this feature is 💯 👌 💅 ✨ for the upcoming `1.0.0` release! Any help is more than welcome, especially with tests on Windows and Linux (and macOS too, to make sure that there are no unexpected side effects) 😄 

Closes #167; Closes #190; Closes #795;

luisrudge commented:
will we be able to run cmder inside of it? http://cmder.net/

AlmirKadric commented:
OH S**T it's finally here
You've got yourself a tester :+1:

factormystic commented:
Windows:

1. running `npm start` for the first time there's no errors, but just a blank black window.

1. After restarting the app, getting an unhandled Exception when starting a new tab (any way; via new tab, new window, split action, etc): 
```
Uncaught Exception:
TypeError: spawn is not a function
    at Session (C:\Users\me\Projects\hyper\app\session.js:40:16)
    at initSession (C:\Users\me\Projects\hyper\app\index.js:377:17)
    at Server.rpc.on (C:\Users\me\Projects\hyper\app\index.js:184:7)
    at emitOne (events.js:96:13)
    at Server.emit (events.js:188:7)
    at Server.ipcListener (C:\Users\me\Projects\hyper\app\rpc.js:34:11)
    at emitTwo (events.js:106:13)
    at EventEmitter.emit (events.js:191:7)
    at WebContents.<anonymous> (C:\Users\me\Projects\hyper\node_modules\electron\dist\resources\electron.asar\browser\api\web-contents.js:210:13)
    at emitTwo (events.js:106:13)
```

1. Maximizing by clicking the maximize button seems to be resizing the window to the size of the active display, but dragging the window still works and doesn't restore the window as it should

1. Snap actions don't work (eg, drag the window to the right side of the screen & release should snap the window to 50% of the display width). Related to this, is that minimizing doesn't animate like normal windows. This is probably due to whatever your electron window configuration is, I believe if you go full borderless you don't get this functionality automatically.

1. The app name & icon when right clicking the task bar icon, and also after pinning the app, is Electron (but maybe this is due to running in development?)

matheuss commented:
@factormystic:

1. Ops, I forgot to write about that on the PR description. 

1.1 Download and install [VC++ Build Tools Technical Preview](http://go.microsoft.com/fwlink/?LinkId=691126)
1.2 Install Python 2.7, add it to your PATH and run `npm config set python python2.7`
1.3 Run `npm config set msvs_version 2015 --global`
1.4 `del node_modules && del app/node_modules`
1.5 `npm install`

This is necessary to build `pty.js` and should fix the issue 😄 I'll add instructions to the PR description and README.md

2./3. I'll look into that 😄 

4. Yeah, this happens when you're running in dev mode – it works as expected when you run the bundled application 😄

matheuss commented:
@luisrudge I don't think so 🤔 From what I understood, `Cmder` is a terminal emulator itself, a full app. Is that correct?

paulcbetts commented:
A few bugs I noticed:

* You need to handle `--squirrel-install` and friends, or else you won't get a shortcut
* Keyboard shortcuts don't seem to work
* If you set your shell to Windows-on-Linux bash (i.e. `C:\Windows\System32\bash.exe`), history doesn't work (i.e. type a command hit enter, then hit up to see it again, it doesn't work)
* Powershell doesn't seem to work as a shell at all
* The default font list doesn't include any Windows fonts

ppot commented:
@matheuss Maybe using pty.js will fix the `TypeError: spawn is not a function` I also receive on my macOS! Strange. Appends sometimes!

supnate commented:
Really cool! But how do you guys make [pty.js](https://github.com/chjj/pty.js) work on Windows!

fhelwanger commented:
It would be nice to have an option on Windows explorer context menu to "Open Hyper here" the same way we can open Atom in some folder:

![image](https://cloud.githubusercontent.com/assets/11396047/19851490/59074ccc-9f45-11e6-8f72-6655947efae4.png)

matheuss commented:
@supnate we're using @Tyriar's [fork](https://github.com/tyriar/pty.js) – all the credit goes to him 😄

Tyriar commented:
Looks great @matheuss 😃, just wanted to let you know some stuff about VS Code's situation so you're all aware:

- This fork has been used in VS Code since the integrated terminal shipped (based on xterm.js and this pty.js fork)
- I consider the state of the fork to be a little messy right now, you can see a summary of the issues here https://github.com/Microsoft/vscode/issues/13625 - if you proceed with this please subscribe to this issue
- As @paulcbetts calls out there are some issues with history doesn't work, there are also some other issues such as ctrl+c not working to kill process, and some other keys not working while in other terminals such as [mongodb and ruby shells](https://github.com/Microsoft/vscode/issues/14613). Another annoying bug is that sometimes the prompt gets out of sync, I haven't identified yet if the cause for this is pty.js or xterm.js
- All Windows pty.js issues for VS Code are being tracked [here](https://github.com/Microsoft/vscode/issues?q=is%3Aopen+is%3Aissue+label%3Aintegrated-terminal+label%3Awindows) which is definitely worth a read. I consider most of these to be artifacts of using an old version of winpty in pty.js
- The actual changes to get it working for VS Code weren't major, but since then there have been some pretty big bug fixes (thanks to @the-ress) https://github.com/Tyriar/pty.js/commits/prebuilt

Despite all these shortcomings, AFAIK it's the best option for all platforms which is why we're using it 😃. It's been on my backlog for a while to spend a sprint to clean up this situation as while many people are happy with the Windows experience, I find it frustrating jumping over from Linux which is so smooth.

jahe commented:
I followed the updated instructions in the pr description and by `npm start` I get an "Uncaught Exception: TypeError: spawn is not a funciton" just like @factormystic:
![exception](https://cloud.githubusercontent.com/assets/2736136/19858040/4369c30a-9f80-11e6-9b7c-47d71eb2654d.png)

cchamberlain commented:
@jahe - What's your `node --version` / `npm -v`?

jahe commented:
@cchamberlain 
`node --version
v6.7.0

npm -v
3.10.3`

Vailfire commented:
Getting the same error as @jahe 
I followed the instructions, but I have Visual Studio 2015 with Visual C++ Tools installed.
```
F:\Projekte\Web\hyper>node --version && npm -v
v6.7.0
3.10.3

F:\Projekte\Web\hyper>npm start

> @ start F:\Projekte\Web\hyper
> concurrently --kill-others --raw "npm run dev" "npm run app"


> @ dev F:\Projekte\Web\hyper
> webpack -w


> @ app F:\Projekte\Web\hyper
> electron app



Webpack is watching the files…

Cannot find module '..\build\Release\pty.node'
A native module failed to load. Typically this means you installed the modules incorrectly.
 Use `scripts/install.sh` to trigger the installation.
 More information: https://github.com/zeit/hyper/issues/72
running in dev mode
electron will open file://F:\Projekte\Web\hyper\app\index.html
Checking for notification messages
ignoring auto updates during dev
No matching notification messages
Hash: fc3fed944d45e1799f8c
Version: webpack 2.1.0-beta.25
Time: 5325ms
           Asset    Size  Chunks             Chunk Names
       bundle.js  4.4 MB       0  [emitted]  main
assets\icons.svg  1.8 kB          [emitted]
    + 317 hidden modules
```

Specro commented:
I also get "spawn is not a function" and during install get errors during child_pty rebuild. Here's a [gist](https://gist.github.com/Specro/8b9b78ed77e172ad43e4839f0e76bcf9).

matthewp commented:
Would you like bug reports here, or does that create too much noise? Would be happy to wait until this lands. I've noticed some issues but probably not serious enough to block merging.

matheuss commented:
@matthewp for sure, feel free to report **everything** here 😄 If you'd like to have a chat, head to [zeit.chat](https://zeit.chat) 😌

matheuss commented:
@jahe @Vailfire I can't reproduce 😔 From a fresh Windows 10 install, I did exactly what I described in the instructions and it was enough to build it :/ 

@Specro from your gist, looks like that you're running on `master`, since you're getting `child_pty` errors. Please make sure that you're on the correct branch 😄

Tyriar commented on appveyor.yml:
```diff
@@ -0,0 +1,31 @@
+  matrix:
+    - platform: x64
+
+image: Visual Studio 2015
```
@matheuss are you sure you need this, [VS Code's appveyor.yml](https://github.com/Microsoft/vscode/blob/master/appveyor.yml) doesn't have anything like it? Having said that, I don't know much about how images are used in appveyor.

matheuss replied:
@Tyriar looks like `Install-Product node 6 x64` [is enough](https://twitter.com/appveyor/status/793176028772302848) 😅 I'll try to remove it later. Thanks for your concern!

matthewp commented:
Ok, here's a few things that i've noticed. With Windows 10 and:

```
> node -v
v7.0.0
> npm -v
3.10.8
```

* Ctrl+t doesn't create a new tab. The file menu works however.
* Using Bash for Windows history doesn't work (I think this is noted elsewhere).
* Paste doesn't work most of the time. I've seen it work some times, but can't figure out a pattern to get it to work. Copy seems to work fine, but that's not used as much (for me at least).

Specro commented:
@matheuss Whoops :D Yup, now it works correctly. I noticed that after some commands the name goes blank on the titlebar. For example, after `npm -v` it changes to npm and after that to nothing. Same with tab titles. Will test more.

jahe commented:
@matheuss I'm running the following system: windows 8.1 pro - 64 Bit - node v6.7.0 - npm 3.10.3 - Python 2.7.12 - Kaspersky 10.2.1.23

I've noticed that the output is slightly different when I run `npm start` in an administrator cmd. The "Uncaught Exception: TypeError: spawn is not a funciton" error occurs either way.

Output without administrator cmd:
```
npm start

> @ start D:\envs\dev\playground\hyper
> concurrently --kill-others --raw "npm run dev" "npm run app"

using node@v6.7.0

> electron app


> @ dev D:\envs\dev\playground\hyper
> webpack -w


Webpack is watching the files…

Hash: fc3fed944d45e1799f8c
Version: webpack 2.1.0-beta.25
Time: 43976ms
           Asset    Size  Chunks             Chunk Names
       bundle.js  4.4 MB       0  [emitted]  main
assets\icons.svg  1.8 kB          [emitted]
    + 317 hidden modules
```

Output with adminstrator cmd:
```
npm start

> @ start D:\envs\dev\playground\hyper
> concurrently --kill-others --raw "npm run dev" "npm run app"


> @ dev D:\envs\dev\playground\hyper
> webpack -w


> @ app D:\envs\dev\playground\hyper
> electron app



Webpack is watching the files…

Cannot find module '..\build\Release\pty.node'
A native module failed to load. Typically this means you installed the modules incorrectly.
 Use `scripts/install.sh` to trigger the installation.
 More information: https://github.com/zeit/hyper/issues/72
running in dev mode
electron will open file://D:\envs\dev\playground\hyper\app\index.html
Checking for notification messages
ignoring auto updates during dev
No matching notification messages
Hash: fc3fed944d45e1799f8c
Version: webpack 2.1.0-beta.25
Time: 48857ms
           Asset    Size  Chunks             Chunk Names
       bundle.js  4.4 MB       0  [emitted]  main
assets\icons.svg  1.8 kB          [emitted]
    + 317 hidden modules
```

Vailfire commented:
After several attempts, I got it finally working for me. I'll try to describe all steps, although I don't know which step fixed the problem exactly.

At first, make sure that your `feature/windows-linux` branch is up-to-date and you deleted your `node_modules` and `app/node_modules` folders.

#### 1. Ensure you have an actual node.js and npm version installed

```
node --version && npm --version
v7.0.0
3.10.8
```

#### 2. Install necessary node_gyp build tools (according to https://github.com/nodejs/node-gyp#readme)

```
npm install --global --production windows-build-tools
```
[Expected output](https://gist.github.com/Vailfire/e533f15b74c9cf1b48cbbee0d9776184#file-2-install-necessary-node_gyp-build-tools)

#### 3. Ensure that your python variable points to the installed executable
I already had python in my PATH variable, pointing to `C:\tools\python2` (like described at https://docs.python.org/2.7/using/windows.html#finding-the-python-executable). The python npm environment variable also pointed to that location. I changed it to point to the actual executable.

```
npm config set python C:\tools\python2\python.exe

python --version
Python 2.7.11
```

(Actual I think the problem was **_here_**.)

#### 4. Install dependencies

```
npm i
```
[Expected output](https://gist.github.com/Vailfire/e533f15b74c9cf1b48cbbee0d9776184#file-4-install-dependencies)

#### 5. Try to build hyper

```
npm start
```
[Expected output](https://gist.github.com/Vailfire/e533f15b74c9cf1b48cbbee0d9776184#file-5-try-to-build-hyper)

#### 6. If it does not work yet?
We can try to compile pty.js to verify that node_gyp is working as expected. Running the windows tests ensures that it is actually **working** when all tests are passed.

```
cd app/node_modules/pty.js

npm i

npm run test-windows
```
[Expected output](https://gist.github.com/Vailfire/e533f15b74c9cf1b48cbbee0d9776184#file-6-if-it-does-not-work-yet)

@jahe @matheuss Hope this can help anyone. 😄

madcampos commented:
@luisrudge:  `cmder` is a standalone app it uses `conemu` as it's "terminal emulator". The other tools provided by it like `clink` should work.

@factormystic @matheuss: window maximizing/minimizing and snapping probably have to do with the way `electron` renders a frameless window similar issues can be found in `nw.js` frameless window. In my last `nw.js` app I solved it in userland code, but the solution is probably outdated now.

Will give it a try and report bugs here. 😉

madcampos commented:
Issues found:
1. Characters like ``'"`~`` get duplicated when inserted in the terminal, like we would expect from a IDE [(image)](https://cloud.githubusercontent.com/assets/609904/19917055/983018ac-a09e-11e6-8261-8dfcd0f93b99.gif)
2. Some characters are rendered as `�` (replacement character) although when run command like `dir` in a Unicode folder it works fine [(image)](https://cloud.githubusercontent.com/assets/609904/19917356/67bfbcba-a0a1-11e6-9790-5a51a3785f66.png)
3. Line height gets too tight when window is resized
4. The contents don't update properly when the window is resized [(image)](https://cloud.githubusercontent.com/assets/609904/19916964/af766a58-a09d-11e6-8d34-5840b8fba2dd.gif)
5. Some emojis don't show (most likely a issue with `cmd` itself) [(image)](https://cloud.githubusercontent.com/assets/609904/19917367/78d14208-a0a1-11e6-8c91-b045b3faeed1.png)
6. Sometimes when the window is maximized than unmaximized the last line gets cut

The _content update_ one is "solved" if i run `vim` and sometimes `cls`, it seems that running something that cleans the screen may solve the issue.

Also I got the following errors when closing the app with the `exit` command.
```
[11864:1102/023137:ERROR:gles2_cmd_decoder.cc(14664)] [.RenderWorker-000001B3F4AE71C0]GL ERROR :GL_INVALID_VALUE : glCopySubTextureCHROMIUM: source texture has no level 0
[11864:1102/023137:ERROR:gles2_cmd_decoder.cc(14664)] [.RenderWorker-000001B3F4AE71C0]GL ERROR :GL_INVALID_VALUE : glCopySubTextureCHROMIUM: source texture has no level 0
[11864:1102/023137:ERROR:gles2_cmd_decoder.cc(14664)] [.RenderWorker-000001B3F4AE71C0]GL ERROR :GL_INVALID_VALUE : glCopySubTextureCHROMIUM: source texture has no level 0
[11864:1102/023137:ERROR:gles2_cmd_decoder.cc(14664)] [.RenderWorker-000001B3F4AE71C0]GL ERROR :GL_INVALID_VALUE : glCopySubTextureCHROMIUM: source texture has no level 0
[11864:1102/023137:ERROR:gles2_cmd_decoder.cc(14664)] [.RenderWorker-000001B3F4AE71C0]GL ERROR :GL_INVALID_VALUE : glCopySubTextureCHROMIUM: source texture has no level 0
[11864:1102/023137:ERROR:gles2_cmd_decoder.cc(14664)] [.RenderWorker-000001B3F4AE71C0]GL ERROR :GL_INVALID_VALUE : glCopySubTextureCHROMIUM: source texture has no level 0
[11864:1102/023137:ERROR:gles2_cmd_decoder.cc(14664)] [.RenderWorker-000001B3F4AE71C0]GL ERROR :GL_INVALID_VALUE : glCopySubTextureCHROMIUM: source texture has no level 0
```

jcolesio commented:
When calling bash from the window. All of the text seems to be highlighted gray. 
![test](https://cloud.githubusercontent.com/assets/4480236/19937622/fac3fa00-a0ef-11e6-81df-47312aadc463.JPG)

zbuttram commented:
Any chance we could get AppVeyor to spit out the packaged version as an artifact? Would love to test but I don't have a machine I'm comfortable installing Python 2 on at the moment.

madcampos commented:
@zbuttram I used [windows build tools](https://github.com/felixrieseberg/windows-build-tools) to build, it installs python and visual c++ compilers in a way that it don't conflict with another installations (I use python 3 by default)

withtobi commented:
Wow looks really nice. The design really fits windows.

zbuttram commented:
Thanks @madcampos, I can also confirm that totally works! Everything seems to be running fine for me. A little slow (like input lag slow), but I'm assuming that's due to running the dev build.

fhelwanger commented:
@jahe I was having the same error using `npm@3.10.3`. Just updated to `3.10.9` and it worked.

Tyriar commented:
On functionality within WSL, there is an upstream issue which causes issues within pty.js (and other ptys). https://github.com/Microsoft/BashOnWindows/issues/111

jahe commented:
@fhelwanger thanks a lot. I updated to node v6.9.1 and npm v3.10.8 and it works like a charme!

jahe commented:
Some things i noticed during test with hyper.exe:

- Closing one of multiple hyper windows causes every hyper window to close instead of just the one being closed
- (Like @Specro already said) Creating multiple Tabs. Initially the title contains "Shell". An "npm -v" causes the title to be blank whereas a simple "dir" command leaves the title "Shell". [image](https://cloud.githubusercontent.com/assets/2736136/19965675/76b82c34-a1c8-11e6-88da-8c990148ac49.png)
- Minimizing the hyper window width and maximizing it again causes the current working directory to be displayed cut [image](https://cloud.githubusercontent.com/assets/2736136/19965767/cb560590-a1c8-11e6-8c9c-18da1a8becb0.png)

jcolesio commented:
Any one having issues running bash.exe besides the arrow keys? Still cant figure out why the text is being highlighted when using bash.exe

flybayer commented:
# Linux Ubuntu 14.04 Feedback

#### As a standalone window next to the default terminal:
![image](https://cloud.githubusercontent.com/assets/8813276/19977837/092e4306-a1c3-11e6-9221-f022ccfc1475.png)

#### Hyper fullscreen:
![image](https://cloud.githubusercontent.com/assets/8813276/19977918/41d891ac-a1c3-11e6-949e-7b01db000606.png)

#### Vs. Gnome Terminal fullscreen:
![image](https://cloud.githubusercontent.com/assets/8813276/19977958/61ce2468-a1c3-11e6-82fe-2f75dcb4281c.png)

#### How the menu appears when fullscreen:
![image](https://cloud.githubusercontent.com/assets/8813276/19978129/db1e149a-a1c3-11e6-8254-186a4bf93c06.png)

## Feedback

1. We need a config option for swapping the location of the hamburger menu and window controls. On some Linux distros, window controls are on the left, and others have them on the right. (_See the mismatch compared to the default terminal in the above top screenshot_)
2. We need a config option for hiding the custom Hyper menu bar when fullscreen. Some Linux distros show the menu bar at the top like OSX (like the bottom screenshot), but others don't, often when the "task bar" is on the bottom of the screen like Microsoft Windows.
3. Something is goofy with the Hyper custom menu elements. I often have to click several times before a click actually registers. I'm not sure if it's due to the size of the click area or something else.

rumpelsepp commented:
The new titlebar with close, maximize and minimize buttons does not make sense in tiling window managers. :/

flybayer commented:
@rumpelsepp I wondered about that. What is best for tiling window managers? Have a config option to hide the entire title bar?

rumpelsepp commented:
> @rumpelsepp I wondered about that. What is best for tiling window managers? Have a config option to hide the entire title bar?

Yep. And maybe let it toggle it with <kbd>alt</kbd> or something.

jcolesio commented:
Noticed that if I start hypertern from Cygwin. I don't get the highlighting issues and I'm able to ssh into machines as needed .

apiep commented:
Tab / Topbar feedback,
can you use chrome tab bar behavior instead of the current implementation?

When unmaximized, it show a little gap between the window border and tab.
![chrome-tab-unmaximized](https://cloud.githubusercontent.com/assets/20897748/20042000/ec6cc334-a4a4-11e6-8a95-d444502ea7a5.png)
And when the window are maximized it will cover the whole area
![chrome-tab-maximized](https://cloud.githubusercontent.com/assets/20897748/20042011/22688e96-a4a5-11e6-8d53-c77f74efeeea.png)

I dont know how hard it is to implement something like that, maybe by adding some empty draggable div below the tab list? :grin:

lednhatkhanh commented:
+1 for @apiep idea, that's far nicer and easier to use than the current implementation!

stefanivic commented:
Just to add another bug i noticed. Setting bash as default shell in Windows trows an error : 
https://i.gyazo.com/c1cb57e35de727cf7b7234cd9ec1369c.png

Tyriar commented:
@stefanivic which bash and what path are you using? If you just used `bash` this is expected.

stefanivic commented:
@Tyriar i tried to use %systemroot%\system32\bash and even to directly point to zsh shell with %systemroot%\system32\bash -c /usr/bin/zsh.

Tyriar commented:
@stefanivic some suggestions:

- Use the extension when referring to the path, it's a shell shortcut to leave out the extension.
- Isn't `%systemroot%` normally the drive, not the Windows directory?
- Also try sysnative over system32, eg. `C:\\Windows\\sysnative\bash.exe`

stefanivic commented:
@Tyriar got it, i didn't use double \\. C:\\Windows\\System32\\bash.exe this is what i ended with. Works now.

jcolesio commented:
@stefanivic 

Do you experience weird highlighting when using bash.exe?
![image](https://cloud.githubusercontent.com/assets/4480236/20111509/3743f512-a5ae-11e6-8565-b9d0caaf28a0.png)

btd commented:
@jcolesio @stefanivic I do not have now, but had because of wrong markup in my PS1 variable.
![default](https://cloud.githubusercontent.com/assets/334851/20111748/74d809e4-a5fa-11e6-8a08-1b58f72a6432.PNG)

jcolesio commented:
@btd 
Thanks. Time to see what mess I have setup.

Tyriar commented:
I would guess that the background highlighting is related to hterm and/or hyper's styles somehow, Haven't seen that in vscode which uses xterm.js.

Specro commented:
I'm interested if there will be someway to change to elevated prompt cause without it things like chocolatey are hard to use. Two more things I noticed:

* cls doesn't clear the screen correctly if there's more text on the screen than there's height:
![cls-bug](http://i.imgur.com/cvIy64k.gif)
* arrow keys don't work in bash nano :/ I guess this is the same bug as bash history thing.

Also if anybody needs these, here's a [gist](https://gist.github.com/Specro/3bbfd94b07f22572aac15862d2fd8f42) of registry entries for menus @fhelwanger described. Just add the path to hyper executable.

RangerMauve commented:
@Specro I'm pretty sure the CLS behavior you're seeing is working the same way it would in the regular windows console.

jcolesio commented:
@Tyriar 

Trying to set back to default to see if that is the issue but this is totally different to the Mac OS version. 

For Windows version, its telling me to edit hyper.js

stefanivic commented:
@jcolesio @btd Nah, no problems here. https://i.gyazo.com/c47360c972a2f81bb6229bc7bf4d9f0c.png

Specro commented:
@RangerMauve I thought so too but when I tried that in the command prompt it clears everything and brings the scrollbar to the top. It doesn't leave any parts of text above.

madcampos commented:
@RangerMauve @Specro I think i saw the same `cls` behavior in `conemu`. IIRC `cls` clears the screen correctly and don't just scroll up since the `cmd` updates on Windows 10, it's worth a try on 7 or 8 to see if there are any differences.

johanpedroo commented:
I can't resolve this problem :/ anyone with this problem? How resolve?
![image](https://cloud.githubusercontent.com/assets/5067128/20118891/7eb6711a-a5ed-11e6-9dba-5a903bb0505f.png)

stefanivic commented:
@johanpedroo Check if you have VC++ Build Tools Technical Preview, also if you already have Visual Studio, just open it and install C++ tools trough it. Also make sure you are on python 2.7.x since node-gyp doesn't support python 3. After that just start all over and it should work.

jcolesio commented:
:( still experiencing issues with bash.exe 

When I start hyperterm from cygwin, then call bash, I dont get that highlighting issues. 

Ideally would like to use WSL bash.exe instead. 

Sooooo close yet soo far.

matheuss commented:
I believe this work achieved a state that it's **acceptable** to live on `master`, so I'm merging it! 
The major issues are solved – Hyper now works on **Windows**! Also, big improvements were made in the **Linux** version.

**Huge thanks** to **everyone** who contributed here! It was amazing to receive that amount of feedback 🙌 

Regarding **arrow keys**/**history** on `bash`: it's upstream as pointed out by @Tyriar 😔 See here for more info: Microsoft/BashOnWindows#111

Now I'll address everything that was reported/suggested here :smile:

***

@factormystic:
 > Maximizing by clicking the maximize button seems to be resizing the window to the size of the active display, but dragging the window still works and doesn't restore the window as it should

I spent some time looking into it and I found that [`window.isMaximized()`] doesn't work – at least on Ubuntu. To be honest, I don't remember if I tested it on Windows – but we chatted about it on [zeit.chat] and we decided to leave as it is for now – since we can't rely on the API. We'd love a Pull Request to fix that 😄 

> Snap actions don't work

I researched about it and found almost nothing 😔 Since it's not a major issue: PRs are more than welcome 😄 

> minimizing doesn't animate like normal windows

Again, found nothing about it. I'm pretty sure it's upstream and there's nothing we can do about it apart from opening an issue on [electron/electron] 😕 

***

@paulcbetts:

> You need to handle `--squirrel-install` and friends, or else you won't get a shortcut

Thank you! It was fixed on https://github.com/zeit/hyper/pull/946/commits/24518b7bc2222a37140aa146e0da14a7d0faabee

> If you set your shell to Windows-on-Linux bash, history doesn't work (i.e. type a command hit enter, then hit up to see it again, it doesn't work)

It's upstream 😕 See Microsoft/BashOnWindows#111 🙌 

> The default font list doesn't include any Windows fonts

cc @CodeTheory @rauchg 

> QuickEdit doesn't work

Since it's not a major issue, I'm leaving it available for an Issue/PR 😄 

> No context menu on right-click

We don't have it on Windows and Linux – so I don't think we'll have it for now on Windows 😅 But it might be useful (on all platforms) – so feel free to open an Issue! ([electron-context-menu] might be useful 😄)

***

@fhelwanger:

> It would be nice to have an option on Windows explorer context menu to "Open Hyper here"

Agreed! It could be very useful on Mac and Linux too. I researched on how they do it on Windows and found [this]. Might be useful. PRs/issues are more than welcome! 😄 

***

@Tyriar:

Thank you for your amazing work on `pty.js`! Without it we wouldn't have Windows support 🙌 
We are aware of the issues – @rauchg gave me a 💯 ✨ to use your fork – we'll eveolve together with [Microsoft/vscode] ✨ 

***

matthewp:

> Paste doesn't work most of the time. I've seen it work some times, but can't figure out a pattern to get it to work. Copy seems to work fine, but that's not used as much (for me at least).

This is not related to Windows – it's related to Hyper. Will be fixed soon! 🙌 

***

@madcampos:

> Characters like '"~ get duplicated when inserted in the terminal [...] The contents don't update properly when the window is resized [...] Sometimes when the window is maximized than unmaximized the last line gets cut [...]

Can't reproduce 😰 

>  Some characters are rendered as � (replacement character) [...] Line height [...]  Some emojis don't show [...]

Both are not specific to Windows. Will be fixed soon! :smile: 

***

@jcolesio:

> All of the text seems to be highlighted gray. 

Can't reproduce :cold_sweat: It's probably something on your setup/env :confused:

***

@jahe:

> Closing one of multiple hyper windows causes every hyper window to close instead of just the one being closed

I noticed this during development and forgot to fix – thanks! Will fix it tomorrow before merging.


> Decreasing the hyper window width and increasing it again causes the current working directory to be displayed cut 

I believe this is not related to Windows – we have it on Mac and Linux too :confused:

***

@Specro and @jahe:

Regarding the window/tab title: thanks for reporting! Turns out that `cmd.exe` likes to report the tab/window title as `' '` :confused: I added a workaround/fix here: https://github.com/zeit/hyper/pull/946/commits/0250aea78ed2e867605d0a5d209e2a2c039fdaaf :smile: 

***

@flybayer:

> We need a config option for swapping the location of the hamburger menu and window controls [...]

**AMAZING** suggestion, thank you! Implemented it on https://github.com/zeit/hyper/pull/946/commits/1867ba81b202fe499defda35323a80d3146b952d :smile: 

> We need a config option for hiding the custom Hyper menu bar when fullscreen

I decided to leave this for later – would love a issue/PR!

> Something is goofy with the Hyper custom menu elements [...] due to the size of the click area [...] 

Thank you again! Fixed on https://github.com/zeit/hyper/pull/946/commits/84993c938b95d5b5b7f0aca186af8d9c48333d06 :raised_hands:

***

@apiep:

> Tab / Topbar feedback [...]

We discussed this before on zeit.chat and we were not able to find a better solution other than the current one :confused: So I decided to leave this for later – would love an issue/PR! :raised_hands:

***

@Specro:

> I'm interested if there will be someway to change to elevated prompt

I'm not sure how it works on Windows – maybe right clicking Hyper and `run as admin` is enough? :thinking:

> cls doesn't clear the screen correctly [...]

Can reproduce; no clue on how to fix :sweat_smile:



[`window.isMaximized()`]: https://github.com/electron/electron/blob/master/docs/api/browser-window.md#winismaximized

[electron/electron]: https://github.com/electron/electron

[electron-context-menu]: https://github.com/sindresorhus/electron-context-menu

[this]: https://github.com/atom/atom/blob/7929e261a0d6e78ff4ca5196c8b385946e64dbd9/src/main-process/win-shell.coffee#L53

[Miscrosoft/vscode]: https://github.com/Microsoft/vscode

Tyriar commented:
@matheuss great news :smile: FYI I enabled issues on https://github.com/Tyriar/pty.js and am working on upgrading winpty and investigating other Windows issues during November/December. I will likely change the name or the repo and publish the module to npm after some improvements are made.

matheuss commented:
Also, forgot to share: 
Regarding the keyboard shortcuts, like `ctrl+n`, `ctrl+t` etc: I'll send a PR later today to fix the i18n stuff – that is, we'll be able to type **any character** with any **keyboard layout** 😄 Such fix involves a good rewrite in the intergration with `hterm` and how focus works, so I'll leave the fix for `ctrl+t` etc for that PR 😄

albinekb commented on lib/components/header.js:
```diff
@@ -81,11 +125,60 @@ export default class Header extends Component {
       onMouseDown={this.handleHeaderMouseDown}
       >
+      {
+        !isMac &&
```
This looks a bit messy, but i guess it works. Maybe break the different cases to new stateless components?

matheuss replied:
Hmm, sounds like a good idea, but I'm not sure how to execute it 🤔 

It'd be something like`<HeaderMac />`?

albinekb commented on app/plugins.js:
```diff
@@ -236,7 +236,7 @@ function install(fn) {
     env.npm_config_disturl = 'https://atom.io/download/atom-shell';
     /* eslint-enable camelcase  */
-    exec('npm prune; npm install --production', {
+    exec('npm prune && npm install --production', {
```
maybe add this as a `script` in package.json?, then `&&` works (even in fish)

albinekb replied:
or run them with .then after each other

```js
exec('npm prune').then(() => exec('npm install'))
```

fearphage replied:
FYI, you wouldn't have to use `&&` in `package.json`.

```json
"pretarget": "npm prune",
"target": "npm install",
```

jcolesio commented:
I wonder if I'm having the highlighting issue due to me using wsltty for the bash console.

madcampos commented:
After update from commit 488e1e8f27545a8aa0486f69bbeb57ac2e09c314 I'm getting the `spawn is not a function` error plus some build errors.

vors commented:
First of all 🎉 
FANTASTIC!

I'm primarily interested in using hyper with `powershell` on windows.
Here is what I found so far

### Copy-paste

I found that `Ctrl+Shift+C` / `Ctrl+Shift+V` works and it's awesome.
PowerShell has it's own way to provide copy-paste with [PSReadLine](https://github.com/lzybkr/PSReadLine).
The handlers logic is implemented in ReadLine. `Ctrl+C` doesn't work and produces `^C` entry, but `Ctrl+V` works just fine. Not sure what's going on, but it would be sweet to make `Ctrl+C` work as well.

### Crashes

When I switch active windows from hyper and wait for some time, I'm getting a reproducible crash

![image](https://cloud.githubusercontent.com/assets/816680/20201612/449cf9f4-a76e-11e6-850b-0a59250e21d6.png)

```
[Window Title]
Error

[Main Instruction]
A JavaScript error occurred in the main process

[Content]
Uncaught Exception:
Error: EPERM: operation not permitted, lstat 'C:\Users\sevoroby\AppData\Local\Temp\notif1161010-13440-14fc7xb.ogn6gmygb9.jpg'
    at Error (native)
    at fs.lstatSync (fs.js:982:18)
    at Object.fs.lstatSync (ELECTRON_ASAR.js:249:16)
    at Gaze._addToWatched (F:\dev\hyper\app\node_modules\gaze\lib\gaze.js:292:14)
    at Gaze._internalAdd (F:\dev\hyper\app\node_modules\gaze\lib\gaze.js:206:10)
    at F:\dev\hyper\app\node_modules\gaze\lib\gaze.js:419:16
    at Array.forEach (native)
    at F:\dev\hyper\app\node_modules\gaze\lib\gaze.js:415:12
    at FSReqWrap.oncomplete (fs.js:123:15)

[OK]
```

### Hyper vs Cmder

Conemu and [Cmder](http://cmder.net/) been on windows for a while, but I didn't use them.
The main reason not to use them was that they slow down native executables. From what I heard, they inject intercept code to do scrapping of stdout and stderr and it adds overhead to all processes.
It slows down my builds about x2 times.

I created a simple one-liner to test it out and I found that hyper doesn't have this overhead problem!
Script that I used:

```powershell
1..200 | % { perl -e 1 }
```

Results

| Terminal  |  Time (sec) |
|----------|--------|
| From powershell.exe: | 4.51|
| cmder |  8.29 |
| hyper | 4.26 |

levrik commented:
I saw on the GIF by @Specro that the scrollbar seems not to have a custom style fitting the style of Hyper. The default windows scrollbar destroys the look and feel a bit.

stefanivic commented:
@levrik I've made an update to the scrollbars. Will push later in the day. 
![](http://i.imgur.com/1dVova4.jpg)

levrik commented:
@stefanivic This looks really nice :o

matheuss commented:
🎉 🎉 🎉 🎉 🎉 🎉

matheuss commented:
@madcampos That commit changed nothing related to the build step, nor `pty.js` 😰 

***

@vors 
Amazing to know that we're that fast ☺️ 
Regarding to that exception: I can't reproduce 😰 
Regarding to copy/paste: I'll introduce a PR very soon that will fix a lot of things related to keyboard issues – including keymaps 😄

vors commented:
@matheuss I don't think that hyper+Powershell can be faster then native host for Powershell, but overhead seems negligible and it's impressive.

levrik commented:
Getting the following error on the version in the `win-dist` folder:

``
Uncaught Exception:
Error: `pty.js` failed to load. Typically this means that it was built incorrectly. Please check the `README.me` to more info.
    at createPtyJsError (D:\Dokumente\Levin\Projekte\Open Source\hyper\dist\win-unpacked\resources\app\session.js:11:32)
    at Object.<anonymous> (D:\Dokumente\Levin\Projekte\Open Source\hyper\dist\win-unpacked\resources\app\session.js:17:9)
    at Module._compile (module.js:556:32)
    at Object.Module._extensions..js (module.js:565:10)
    at Module.load (module.js:473:32)
    at tryModuleLoad (module.js:432:12)
    at Function.Module._load (module.js:424:3)
    at Module.require (module.js:483:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (D:\Dokumente\Levin\Projekte\Open Source\hyper\dist\win-unpacked\resources\app\index.js:30:17)
``

vors commented:
@matheuss actually my statement about **reproducible crash** was wrong. I got it repro twice and could not repro it anymore as well.

levrik commented:
@factormystic That the snap feature of windows is not working and the minimize/restore animations are not there is caused by the `transparant` setting set on the window. More information here: http://electron.atom.io/docs/api/frameless-window/#transparent-window

aleksijohansson commented:
Really looking forward to this getting released!

jasonmit commented:
@jcolesio I also ran into this.  The fix was to just delete HKEY_CURRENT_USER/Console in the registry.

But the core issue is the color palette at the system level is partially leaking into hyper.  Should I file a bug for this @matheuss and friends?

Great work by the way :)

jcolesio commented:
@jasonmit  You da man!!!!!!!! worked !

jasonmit commented:
@jcolesio https://github.com/Microsoft/BashOnWindows/issues/111 looks to be the source of the the arrow key issue

DanielPintilei commented:
`transparent: false` BrowserWindow default would be nice; With `transparent: true` the window flickers sometimes on `win.show()`

brighthero commented:
When will there be a release windows version?

ppot commented:
in the 1.0.0 version. Mostly in the next commings weeks. @matheuss

howdoicomputer commented:
Oh my God, this is amazing. As a heavy user of cmder, I can't be happier that terminal innovation is coming to Windows.

thekalinga commented:
Its not working on Ubuntu 14.04.

1. `Ctrl+A` used to take me to the first character in the terminal before
=> Now it does all text in the window like a browser does

2. Select text & try right clicking/pressing `Ctrl+C`, the selection is gone.
=> Text cannot be copied

ppot commented:
@thekalinga  2. On windows you need to use : `ctrl+shift+c` for copy.
see : https://github.com/zeit/hyper/blob/master/app/accelerators.js
-> Right clicking not implemented right now. 
1. `ctrl+A` is used to selectAll like any linux terminal. 
-> First character on the terminal top or bottom.

plessbd commented:
@ppot 
> ctrl+A is used to selectAll like any linux terminal.
-> First character on the terminal top or bottom.

I believe most terminals ctrl+a moves to the beginning of the line and ctrl+e moves to the end of a line

see:
http://ss64.com/osx/syntax-bashkeyboard.html
http://lifehacker.com/5743814/become-a-command-line-ninja-with-these-time-saving-shortcuts
http://www.howtogeek.com/howto/ubuntu/keyboard-shortcuts-for-bash-command-shell-for-ubuntu-debian-suse-redhat-linux-etc/

trodrigues commented:
I just landed here looking for info on how to run Hyper with Bash on windows and found this [comment](https://github.com/zeit/hyper/pull/946#issuecomment-259209300) by @Tyriar.

I wonder if there's a good place to document this to prevent someone getting so lost in the future. I don't see an FAQ section in the website.

Happy to provide a PR for this but wondering where the best place is.

webframp commented:
As mentioned by @plessbd:

> @ppot 
> ctrl+A is used to selectAll like any linux terminal.
> -> First character on the terminal top or bottom.
> I believe most terminals ctrl+a moves to the beginning of the line and ctrl+e moves to the end of a line

It seems like a bad idea to override any gnu readline default keybindings without some way for the user to disable as the majority of shells (bash, zsh, fish) will implement basic readline bindings by default. I definitely prefer `ctrl+a` to move to the start of the line, not select all.

ppot commented:
@webframp `cmd+a` is selectAll in macOS. `ctrl+a` is also true for what you are saying. The key-mapping need to be changed.

ppot commented:
@webframp  hterm interpret those actions as: `ctrl+a === ctrl+shit+a  ` and `ctrl+e === ctrl+shit+e`

webframp commented:
Thanks @ppot, ctrl+shift+a if not already taken would make sense as select-all and still preserves ctrl+a start of line behavior

lasermarty commented:
Any news on 1.0.0?

fhelwanger commented:
It's already released, 1.1.0 is the latest version: https://github.com/zeit/hyper/releases/tag/1.1.0

katacarbix commented:
I would like to see support for Bash on Windows (WSL). Maybe a setting in `.hyper.js`?
Thanks.

matheuss commented:
@katacarbix it's just a matter of putting `bash` on `shell` on `.hyper.js`. 

PS: If it doesn't work, you need the absolute location

AlmirKadric commented:
@katacarbix also take a look at this PR #1347
it provides the ability to support multiple shells from the `.hyper.js` config

sabahang commented:
@jasonmit @jcolesio had the same highlighting issue. It was one of the following registry keys that did the trick:
"PopupColors"=dword:000000f5
"ScreenColors"=dword:00000087
"EnableColorSelection"=dword:00000000

Not sure which one did the trick. For other running into same issue try one by one until it works! no restart needed!

19lmyers commented:
Why can't you just flag the buttons as non-draggable on the old design, so they can receive click events? 
The main thing irritating me about this app right now is that the tab bar is below the title bar, an obvious regression from the Mac OS version.

ppot commented:
Insted of commenting on this PR, please open new issue and it will be discussed there. 
Thanks! 💪
